### ISSUE TITLE
Java: Issue with ``virtual`` as identifier

### ISSUE THREAD

**Issue by Stewori** [doxygen/doxygen#8882]
**Describe the bug**
Processing Java code
```
public class Test {
    public boolean virtual = false;
}
```
results in the following xml snippet:
```
<memberdef kind="variable" id="classTest_1a0448d2bed34ae7e57205e14d74edab6f" prot="public" static="no" mutable="no">
 <type></type>
 <definition>virtual Test.boolean</definition>
 <argsstring></argsstring>
 <name>boolean</name>
 <initializer>= false</initializer>
 <briefdescription>
 </briefdescription>
 ...
</memberdef>
```
Note the wrong name and missing type. The issue does not occur if there is no initializer. I suspect doxygen is confused by the identifiier ``virtual`` which is a reserved keyword in C++ but not in Java.

**Expected behavior**
I expect that the resulting code snippet would be
```
<memberdef kind="variable" id="classTest_1a0448d2bed34ae7e57205e14d74edab6f" prot="public" static="no" mutable="no">
 <type>boolean</type>
 <definition>boolean Test.virtual</definition>
 <argsstring></argsstring>
 <name>virtual</name>
 <initializer>= false</initializer>
 <briefdescription>
 </briefdescription>
 ...
</memberdef>
```

**To Reproduce**
Find a Java file and doxygen config attached. The issue is demonstrated like mentioned above.
After running doxygen on the example, the files in the subfolder ``xml`` contain the snippet shown above.
[issue_Java_virtual.zip](https://github.com/doxygen/doxygen/files/7499727/issue_Java_virtual.zip)

**Version**
Tested with doxygen clone from 2021-11-08:
```
/blah/doxygen/build/bin/doxygen -v
1.9.3 (096a1c10e1e0b86edc8c561f49b5fd7881776ba7)
```

`lsb_release -a`:
```
LSB Version:	core-9.20160110ubuntu0.2-amd64:core-9.20160110ubuntu0.2-noarch:security-9.20160110ubuntu0.2-amd64:security-9.20160110ubuntu0.2-noarch
Distributor ID:	LinuxMint
Description:	Linux Mint 18 Sarah
Release:	18
Codename:	sarah
```

**Additional context**
TBH I did not check for duplicates. Please excuse if this is a known issue.

**Comment by Stewori** [comment]
It occurred to me that maybe other C++ keywords that are not keywords in Java (e.g. ``inline``) might provoke the issue similarly.

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #8886 which fixes the problem for `inline`, `requires` and `virtual`.
Maybe there are more words  ..., so a small example and its is easily added to the proposed pull request (as long as the proposed pull request is not integrated into master, when integrated than we need a new issue / pull request).

I don't think it is wise to use such normal words as identifiers, but sometimes it does make sense.
The might indeed be more keywords that should be ignored for Java, I had a quick look in `code.l` as here there is a bit of a list:
```
KEYWORD_OBJC ("@public"|"@private"|"@protected"|"@class"|"@implementation"|"@interface"|"@end"|"@selector"|"@protocol"|"@optional"|"@required"|"@throw"|"@synthesize"|"@property")
KEYWORD ("asm"|"__assume"|"auto"|"class"|"const"|"delete"|"enum"|"explicit"|"extern"|"false"|"friend"|"gcnew"|"gcroot"|"set"|"get"|"inline"|"internal"|"mutable"|"namespace"|"new"|"null"|"nullptr"|"override"|"operator"|"pin_ptr"|"private"|"protected"|"public"|"raise"|"register"|"remove"|"self"|"sizeof"|"static"|"struct"|"__super"|"function"|"template"|"generic"|"this"|"true"|"typedef"|"typeid"|"typename"|"union"|"using"|"virtual"|"volatile"|"abstract"|"final"|"import"|"synchronized"|"transient"|"alignas"|"alignof"|"concept"|"requires"|"decltype"|{KEYWORD_OBJC}|"constexpr"|"consteval"|"constinit"|"co_await"|"co_return"|"co_yield"|"static_assert"|"noexcept"|"thread_local")
FLOWKW  ("break"|"catch"|"continue"|"default"|"do"|"else"|"finally"|"return"|"switch"|"throw"|"throws"|"@catch"|"@finally")
FLOWCONDITION  ("case"|"for"|"foreach"|"for each"|"goto"|"if"|"try"|"while"|"@try")
TYPEKW  ("bool"|"byte"|"char"|"double"|"float"|"int"|"long"|"object"|"short"|"signed"|"unsigned"|"void"|"wchar_t"|"size_t"|"boolean"|"id"|"SEL"|"string"|"nullptr")
TYPEKWSL ("LocalObject"|"Object"|"Value")
CASTKW ("const_cast"|"dynamic_cast"|"reinterpret_cast"|"static_cast")
```
the problem also exists in `scanner.l` but is a bit more work here.

I've just pushed a proposed patch, pull request #8886 which fixes the problem for `inline`, `requires` and `virtual`.
Maybe there are more words  ..., so a small example and its is easily added to the proposed pull request (as long as the proposed pull request is not integrated into master, when integrated than we need a new issue / pull request).

**Comment by Stewori** [comment]
It should be doable to lookup "somewhere" authoritative lists of reserved keywords for Java, C++ and other languages.
I can check this out for C++ and Java (maybe for Python too) to enable a reliable fix here. However for other languages someone else has to care (atm).
Would it help if I provided such lists for you here?

**Comment by albert-github** [comment]
That would be nice. Python is probably not a problem as the python scanners are independent of the C++ / Java / C# scanners, so no overlap.
- I think a good sources for C++ are the standard but also https://en.cppreference.com/w/cpp/keyword

**Comment by Stewori** [comment]
For Java, it's here: https://docs.oracle.com/javase/tutorial/java/nutsandbolts/_keywords.html
However I still look for something explicitly at current Java 17 state.

**Comment by Stewori** [comment]
Okay, for Java 17 the very authoritative list is here: https://docs.oracle.com/javase/specs/jls/se17/html/jls-3.html#jls-3.9
However, it got some new keywords that may cause problems in much current Java code, e.g. ``var``.
There are these "contextual keywords" now, which as I understand it mostly apply just to module-info files. But e.g. ``sealed`` and ``non-sealed`` can apply to class declarations (?). I think this has to do with recently introduced records but I am not sure. Anyway, this appears to be a wider issue than expected. Do you have plans in what manner to support Java 17? Maybe a new flag is required to opt in or out of a certain modern set of Java keywords.

- adopting the new set of keywords would cause issues on vast amounts of current Java code where these may occur as identifiers
- not adopting the set of new keywords may cause issues for Java 17 which is the future after all

OTOH it never was such a big problem that Doxygen does not recognize Java keywords such as ``volatile``, ``transient``, etc. so I think the best way to go would to recognize as few Java keywords as possible. Maybe the Java 8-level list linked in the comment before would be a good choice.
After all, this issue is caused by recognizing too many keywords (``requires``, ``virtual``, etc), not the other way round.

**Comment by albert-github** [comment]
Thanks for the lists.
I'm not sure what the best strategy would be. I have to dive into it a little bit.
I did see some keywords I don't like: `_` and `non-sealed`, i.e. a keyword of being just an underscore and for `non-sealed` quite easily confused with the minus sign.

**Comment by Stewori** [comment]
I think these won't cause trouble because neither a sole underscore nor words containing the minus sign have ever been allowed as Java identifiers.

**Comment by Stewori** [comment]
> I think a good sources for C++ are the standard but also https://en.cppreference.com/w/cpp/keyword

That's quite a lot keywords. Doxygen does not recognize all of them, does it?

Maybe we should just leave this issue with your current solution. If I find more issues I will report them as they pop up.

**Comment by albert-github** [comment]
I think doxygen support most of the keywords (except when the are not too new or when they are not necessary, I don't see e.g. `xor` somewhere in the C++ scanners).

I think that when we can easily add the keywords like I did for `inline`, `requires` and `virtual`  I think this should be done now as we have now the momentum.

**Comment by Stewori** [comment]
Allright. Based on the linked lists and some practicability considerations I would suggest that the following list of C++ keywords should be considered valid Java identifiers:
```
alignas
alignof
and
and_eq
asm
atomic_cancel
atomic_commit
atomic_noexcept
auto
bitand
bitor
bool
char8_t
char16_t
char32_t
compl
concept
consteval
constexpr
constinit
const_cast
co_await
co_return
co_yield
decltype
delete
dynamic_cast
explicit
export
extern
friend
inline
mutable
namespace
noexcept
not
not_eq
nullptr
operator
or
or_eq
reflexpr
register
reinterpret_cast
requires
signed
sizeof
static_assert
static_cast
struct
template
thread_local
typedef
typeid
typename
union
unsigned
using
virtual
wchar_t
xor
xor_eq
```

**Comment by Stewori** [comment]
(And we should keep in mind to test doxygen with some Java example that contains ``non-sealed`` or ``_``, but that's orthogonal to this issue.)

**Comment by albert-github** [comment]
Thanks for the lists, I just pushed an extension to my previous proposed pull request.

From the second and third list I I think only `override` give a code coloring problem (so I added it as well), but when you have an example proving the opposite ...

**Comment by Stewori** [comment]
I mostly discover these issues by the sanity check that ``<type></type>`` should not be empty except for constructors and ``[static/instance initializer]``s. If I get an alert again I will file an issue.

**Comment by Stewori** [comment]
I confirm that your PR fixes the issue, at least for my setting.

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.3.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
